src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -119,69 +102,69 @@ private void Start(string workingDirectory, string fileName, string arguments, I
 
             // Hook up the events.
             _proc.EnableRaisingEvents = true;
-            _proc.Exited += OnExited;
-            if (_proc.StartInfo.RedirectStandardOutput)
-            {
-                _proc.OutputDataReceived += OnOutputDataReceived;
-            }
+            _proc.Exited += ProcessExitedHandler;
 
-            if (_proc.StartInfo.RedirectStandardError)
+            using (var registration = cancellationToken.Register(() => CancelProcessTree()))
             {
-                _proc.ErrorDataReceived += OnErrorDataReceived;
-            }
+                // Start the process.
+                try
+                {
+                    _stopWatch = Stopwatch.StartNew();
+                    _proc.Start();
+                }
+                catch (Exception ex) when (ex is Win32Exception && (ex as Win32Exception)?.NativeErrorCode == 2)
+                {
+                    throw new FileNotFoundException($"File not found: {_proc.StartInfo.FileName}", ex);
+                }
 
-            // Start the process.
-            _stopWatch = Stopwatch.StartNew();
-            bool newProcessStarted = _proc.Start();
-            if (!newProcessStarted)
-            {
-                Trace.Verbose("Used existing process instead of starting new one for " + fileName);
-            }
+                // Close the input stream. This is done to prevent commands from blocking the build waiting for input from the user.
+                _proc.StandardInput.Dispose();
 
-            // Start reading output.
-            if (_proc.StartInfo.RedirectStandardOutput)
-            {
-                _proc.BeginOutputReadLine();
-            }
+                // Start the standard error notifications, if appropriate.
+                if (_proc.StartInfo.RedirectStandardError)
```

Comment:
I don't like that both STDOUT and STDERR should be processed asynchronously. It seems like this approach is prone to process the data out of sequence.

I think only STDOUT should be processed in this asynchronous fashion. For STDERR, the regular synchronous event handler should be hooked up. In that synchronous event handler for STDERR: check if STDOUT has more data (if so process it first), then process the line from STDERR.

That would guarantee the order of STDERR/STDOUT for programs that write synchronously to STDERR.